### ISSUE TITLE
Grandia II : DSP Audio causes too much lag. Seperate issue: some Graphical floor glitches

### ISSUE THREAD

**Issue by DreaMBlazeR** [flyinghead/flycast#327]
**Please Note: This form is the minimum required information for submitting bugs.**  
**Removing this form may lead to your issue being closed until it is completed.**

Platform / OS / Hardware: Lenovo A7-3300/GV Android kitkat 4.4

Github hash: 

Game: GRANDIA 2 NTSC USA GDI

Hardware: Lenovo A7-3300/GV

**_Description of the Issue_**
Bugs:
1. Using DSP Audio causes too much lag/ if off, SFX volumes like dubbed voices/hallway echoes/magic sfx/ go way high in intolerable db and BGM is too low to be heard.
2. Graphical glitches on floors when in town.
3. Flycast standalone crash after fighting Tongue of Valmar and then going into the INN to save.
4. Some minor icon item display glitch whenever being obtained the first time.

**_Debugging Steps Tested_**

1. How to fix SFX issue:

(DSP bug present on Flycast standalone emulator can be fixed by using Retroarch with NAOMI Flycast App instead)
*Turning it off while in-game will make the game fps go full speed
*Turning it on again while in-game will erase the "volume dub/Loud SFX" fix but fps goes half(30-20fps) allow 5 secs to take effect, then
*Turning it off again while in-game will retain the full fps speed back and the loud unbearable SFX (like hallway echoes/fountains/voice dubs) will be fixed. BGM can now be heard. I tried frame rate skipping but it just slows down the game, DSP toggling is my only way to make the the SFX bearable.

*So everytime you start e game, make sure DSP is off toggle it while in game, run the game, then go back to menu and leave it off. This toggle trick is also useful on the OP FMV so that the music won't stutter.

2. Graphical glitches on floors can not be fixed at this time, looking for suggestions. I already tried delay frame swapping, resolution changing even using 2xBXR filter, nothing works.
3. Always have back-up saves (not save-states). If flycast standalone emulator crashes on your way going to the inn, use alternate save then try the game on retroarch app. Please note, RA has some bloatware issue as of the moment (you need 450+MB) just to run a DC/NAOMI Atomisware game. They said they will fix this soon.
I was also using a 2xBRX filter when I was passing into the INN after the fight with the tongue of valmar.
4. Some newly obtained Icons display junk text when Delay Frame Swapping is off, so turn it on.

**_Logs Gathered_**

```

Please paste the contents of the log / logs here
You may leave this field blank if you have none.

```

**_Screenshots_**

(Replace this line with any issue screenshots)
![Grandia II v1 001 (2000)(Ubi Soft)(NTSC)(US) ! -210823-191852](https://user-images.githubusercontent.com/89394472/130452630-397e32bb-644c-46df-a6ac-92fb93ce2796.png)
![Grandia II v1 001 (2000)(Ubi Soft)(NTSC)(US) ! -210823-191925](https://user-images.githubusercontent.com/89394472/130452694-a98cdc19-4f37-45c8-afb7-0eeb96d378c5.png)

**Comment by flyinghead** [comment]
Can you attach a vmu save or a savestate so I can reproduce the Tongue of Valmar crash?

Also which standalone version (git hash) are you running?

**Comment by satchnut** [comment]
Haha DreaMBlazeR beat me to it. 

Confirmed issue spanning old releases up to the latest nightly IPA build of Flycast. Both Japanese/English versions of the game are affected.

**Comment by DreaMBlazeR** [comment]
tsk, I forgot to retain the old save, sorry. Last github build I was using was this,
![Screenshot_2021-08-24-13-50-17](https://user-images.githubusercontent.com/89394472/130562589-c445f487-df7b-4469-b7fc-d7adbe9882c8.png)

**Comment by Gamemulatorer** [comment]
> tsk, I forgot to retain the old save, sorry. Last github build I was using was this,
> ![Screenshot_2021-08-24-13-50-17](https://user-images.githubusercontent.com/89394472/130562589-c445f487-df7b-4469-b7fc-d7adbe9882c8.png)

Because your phone is too weak for DSP
![Screenshot_2021-08-24-13-47-27-275_com flycast emulator](https://user-images.githubusercontent.com/37603562/130562895-e68a70d0-a0ed-43ef-8ab2-7f979bb1d78f.jpg)
That should only be on if your phone is 64bit not 32bit

**Comment by DreaMBlazeR** [comment]
Yeah I guess my QuadCore can't handle it. Capcom vs SNK 2 and Marvel vs Capcom which are 2D games run well even if it is on. Grandia 2 which is a demanding 3D game has a lot of audio tracks, (4 of them), problem is when DSP is off the SFX and voices goes through the roof, it is very painful in the ear even at 35% volume. Problem does not exist in libreto when using toggle method. I wonder what causes the audio bug. When I looked at the steam Anniversary edition, on forums they also spoke of such SFX bug. On a separate note, the floor glistening bug is something I could not fix. Guess I'll be forced to use Retroarch. I like Flycast more cause whenever I exit the RA App, there are garbage icons replacing visual buttons whenever I resume it. Also Flycast only requires 50MB.

**Comment by DreaMBlazeR** [comment]
Just look at the difference.
![IMG_20210824_142448_078](https://user-images.githubusercontent.com/89394472/130566160-2f05ce68-9cbc-4dbf-bd25-170c1865c25a.jpg)

**Comment by DamnBoi706** [comment]
it's like the the game rely heavily on Dreamcast DSP processor for audio so disable DSP emulation would result either glitchy sound or no sound at all but emulate the Dreamcast DSP would require some additional Power which explain why it only recommended for Fast devices

**Comment by DamnBoi706** [comment]
And those graphical issue maybe caused by open GL ES 2.0 instead of 3.0 i guess? again don't take my word on that I'm just guessing

**Comment by DreaMBlazeR** [comment]
https://drive.google.com/file/d/1VmTzoppOHXzeRpeQOHT_ESkzR-aPJ-4S/view?usp=drivesdk

https://drive.google.com/file/d/16GMIEijuDt-ZzY8zq9Aewap5I70yzXIg/view?usp=drivesdk

**Comment by flyinghead** [comment]
The savestate crashes almost instantly so I assume it's already corrupted.

The vmu save you posted is empty (D1) but I was able to recover the vmu (A1) from the savestate. It has 3 saves:
![image](https://user-images.githubusercontent.com/16261910/130587710-cdea36d9-3927-4ba9-a55e-e9760b76b6a0.png)

Can you tell me how to reproduce the crash starting with any of these saves?

**Comment by DreaMBlazeR** [comment]
Load Liligue cave 3 save file (7:12:48) go inside the cliff, fight Tongue of Valmar. After the boss fight, a scene plays where Gadan is incapacitated, then when you are able to control the character try to go to the inn to save (on your way there, just outside the INN building it will crash). The save state is when I already defeated the boss, and when I am already near the INN.

**Comment by flyinghead** [comment]
This is what I did but it didn't crash for me: after defeating Tongue of Valmar, I was able to enter the INN and save the game. (Android, arm32, same game version)

Can you attach your emu.cfg so I can replicate your settings?

**Comment by DreaMBlazeR** [comment]
https://drive.google.com/file/d/1i48Pxs8vjNL8J5l3U114Lb92CBNCBVHu/view?usp=drivesdk

emu.cfg

**Comment by flyinghead** [comment]
Still no luck, this time with your same settings. (Note that you shouldn't use Texture Upscale (xBRZ) on such a slow device. It's set to x2 in your configuration, although not for Grandia 2.)

I suggest you install the latest build from the libretro branch, which is the current 'dev' branch. It has a crash detector that will create a `.dmp` file in case of crash. You can then attach the dump file to this issue for analysis.

Regarding the other issues:

* There is no arm32 dynarec for the DSP, which explains why it's slow. I'll see if I can fix some audio issues when the DSP is off.
* OpenGL ES 2.0 is limited and Mali 400 is terrible with float precision so there are graphic glitches and there isn't much I can do. GLES 3.0 is preferred.

**Comment by flyinghead** [comment]
I added an arm32 DSP dynarec implementation and also fixed a problem with DSP sounds being too loud when the DSP is disabled. It should improve things both when the DSP is on and off.

These changes are in the latest build of the libretro branch.

**Comment by DreaMBlazeR** [comment]
Recent build of retroarch(libreto) fixed graphical issues on layering on floors. I was in Saint Granas Cathedral and floor isn't glistening with shadows anymore. DSP Audio still needs to be toggled (on and off) while in game to hear BGM and SFX correctly. So far everything looks polished. Please note, crash only involves flycast standalone app not the core in RA. Not encountering anymore bugs/crash in RA  the moment. Thanks a lot for devs for fixing the emulation system.
![Grandia II v1 001 (2000)(Ubi Soft)(NTSC)(US) ! -210825-152354](https://user-images.githubusercontent.com/89394472/130750807-2a102271-209b-4edf-a47d-0d0f74dfed91.png)
![Grandia II v1 001 (2000)(Ubi Soft)(NTSC)(US) ! -210825-152427](https://user-images.githubusercontent.com/89394472/130750855-79dde5a3-562c-4430-835f-bf975955724a.png)

**Comment by flyinghead** [comment]
Can you test the latest standalone build on the libretro branch? You can find it on the builds page: https://flyinghead.github.io/flycast-builds/.

The libretro branch has both standalone builds AND libretro core builds. Sorry for the confusion.

**Comment by DreaMBlazeR** [comment]
Dunno why, but for some reason the standalone and libreto version seems slower than the RA build. On the new requested flycast build there are no more crashes, I turned off Vsync, Delay Frame Swapping, DSP, but scrolling through towns have frame rate drops to 25-20ps instead of constant 30fps on RA. Also first time I got audio stuttering on towns, not existent on RA. SFX and loud vocals are fixed though. No graphical glitches on St. Heim Papal state, but the LILI INN still has their roads glistening. I'll try to run it again when fully charged, RA has a feature called "sustained performance mode" that makes itself fun at full speed even at low power. Probably report later sorry.
